### ISSUE TITLE
run-vcpkg post-action doesn't run after failed step

### ISSUE THREAD

**Issue by stevenlovegrove** [lukka/run-vcpkg#58]
I have a very long running vcpkg install phase and had a CMake error in a step after. This seemed to cause the post-action step of run-vcpkg to not get executed and hence not cache the installed packages. This meant that attempts to fix the CMake error resulted in long test cycles since the same packages still had to get built each time and were never cached. Since the git checkout post-action still seems to run, I guess it is possible to perform these post-actions even on failure. Looking through the run-vcpkg code, I couldn't figure out what was stopping it (but I'm not experienced with github actions).

Do you agree that it would be sensible to cache the installed packages provided the cache install step succeeded? Am I using run-vcpkg incorrectly (https://github.com/stevenlovegrove/Pangolin/actions/runs/406547331/workflow#L40 in case seeing the workflow helps)?

Thanks in advance!

<img width="404" alt="Screen Shot 2020-12-07 at 10 08 56 PM" src="https://user-images.githubusercontent.com/481156/101447290-c63edf80-38d9-11eb-8231-a17772b4ccdf.png">

**Comment by vadi2** [comment]
Update vcpkg to latest, this behaviour has been improved upon.

**Comment by lukka** [comment]
@stevenlovegrove at the moment `run-vcpkg` caches the artifacts when the step succeeded, but this does not happen when `setupOnly:true`, where it happens as a post action and only if the workflow succeeded. It makes definitively sense to have the same logic whatever the value of `setupOnly` is indeed.

Regarding your workflow, this [build.yml](https://github.com/lukka/Pangolin/blob/use_vcpkgjson_and_run-cmake/.github/workflows/build.yml) is how I would rewrite it: 
 - based on `vcpkg` for all platforms. In an ideal world there would be a multiplatform package manager for C++, in real life having something installed via the platform package manager (ie. apt, yum, pack, brew) is still needed for some dependencies.
 - specifying dependencies in `vcpkg.json`(similar to `npm`'s `packages.json`, but not as powerful yet).
 - using [run-cmake](https://github.com/lukka/run-cmake) action to install deps, generate projects and build with CMake and Ninja (which would be faster on all platforms.)

**Comment by stevenlovegrove** [comment]
Thanks for the info! And I really appreciate the script change example, thanks @lukka!

**Comment by lukka** [comment]
@stevenlovegrove  note that the workflow as is does not work yet. Last run is [here](https://github.com/lukka/Pangolin/runs/1518108154?check_suite_focus=true)

The dependencies needs to figured out, but once that is done, it should be just working fine. 

If you can take a look and tell me what errors I am doing, I am willing to get to the bottom of any other remaining problems and create a PR upstream with the changes.  
Of course, I am open to any changes may be required, nothing in this workflow is definitive, all can be changed

**Comment by stevenlovegrove** [comment]
I think I will keep a similar structure as I have for now because I am attempting to keep the process very similar to the user install, configure, build sequence. The main reason I don't want to use vcpkg for all platforms is because I want to recreate the build environment of my users, and vcpkg is not yet so popular on Linux or Mac. That way I can hopefully hit the problems users will encounter before they do. If it were easy to add an option for run-vcpkg that would cache even on failure, I'd be happy to try to modify it myself if you perhaps can give a pointer?

**Comment by lukka** [comment]
@stevenlovegrove to cache vcpkg's artifacts even on workflow failure try ` uses: lukka/run-vcpkg@dev/always_cache_in_post_action`.  That is what I used the [build.yml](https://github.com/lukka/Pangolin/blob/use_vcpkgjson_and_run-cmake/.github/workflows/build.yml#L43).

**Comment by stevenlovegrove** [comment]
Ah! I missed that, thanks!

**Comment by lukka** [comment]
Please take a look to [run-vcpkg@v10](https://github.com/marketplace/actions/run-vcpkg) and see if suites your needs. Feel free to reopen this or a new ticket about any problem you may see.
